### PR TITLE
Feature/followed

### PR DESCRIPTION
**Pull request by Keith-Go**
## Ticket 
- [feature followed](https://trello.com/c/9L659HXZ/8-followed-event)

## To do
- List event upcoming
- List event past

## Video
[record.webm](https://github.com/luke-nguyen-goldenowl/join_me/assets/152830147/ef9b2147-889b-468e-8fd6-0993018ea587)

### CODE REVIEW

File: lib/src/features/followed/past/logic/past_bloc.dart [hunk 1]
```diff
@@ -9,10 +12,35 @@ import 'package:myapp/src/network/model/event/event.dart';
...
+
   @override
-  Future<MResult<MPaginationResponse>> get getDataAPI async {
-    await Future.delayed(const Duration(seconds: 2));
+  Future<MResult<MPaginationResponse<MEvent>>> get getDataAPI async {
+    final data =
+        await domain.event.getEventsPastByUser(user.id, state.data.lastDoc);
 
-    return domain.eventMock.getEventsByUser('1');
+    if (data.isSuccess) {
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/followed/past/logic/past_bloc.dart]: Sao chỗ này giống đang mock data v em, mình đang gọi trong Bloc mà, nếu mock data thì phải handle ở repository chứ

File: lib/src/features/followed/past/view/past_event_followed_view.dart [hunk 2]
```diff
@@ -10,35 +10,36 @@ class PastEventFollowedView extends StatelessWidget {
...
-                    ),
-                  );
-                } else {
-                  return Ticket(
-                    event: state.data.data[index],
-                  );
-                }
-              }),
+    return BlocBuilder<PastBloc, PastState>(
+      buildWhen: (previous, current) => previous != current,
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/followed/past/view/past_event_followed_view.dart]: update buildWhen

File: lib/src/features/followed/upcoming/logic/upcoming_bloc.dart [hunk 3]
```diff
@@ -1,18 +1,46 @@
...
   @override
-  Future<MResult<MPaginationResponse>> get getDataAPI async {
-    await Future.delayed(const Duration(seconds: 2));
+  Future<MResult<MPaginationResponse<MEvent>>> get getDataAPI async {
+    final data =
+        await domain.event.getEventsUpcomingByUser(user.id, state.data.lastDoc);
 
-    return domain.eventMock.getEventsByUser('1');
+    if (data.isSuccess) {
+      final MPaginationMeta meta;
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/followed/upcoming/logic/upcoming_bloc.dart]: Use mock in the repository

File: lib/src/features/followed/upcoming/view/upcoming_event_followed_view.dart [hunk 4]
```diff
@@ -10,35 +10,36 @@ class UpcomingEventFollowedView extends StatelessWidget {
...
-                    ),
-                  );
-                } else {
-                  return Ticket(
-                    event: state.data.data[index],
-                  );
-                }
-              }),
+    return BlocBuilder<UpcomingBloc, UpComingState>(
+      buildWhen: (previous, current) => previous != current,
```
Comment (luke-nguyen-goldenowl) [inline lib/src/features/followed/upcoming/view/upcoming_event_followed_view.dart]: update buildWhen

File: lib/src/network/data/event/event_reference.dart [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+
+class EventReference extends BaseCollectionReference<MEvent> {
+  EventReference()
+      : super(
+          FirebaseFirestore.instance.collection('events').withConverter<MEvent>(
+                fromFirestore: (snapshot, options) => MEvent.fromMap(
+                    snapshot.data() as Map<String, dynamic>, snapshot.id),
+                toFirestore: (chatRoom, _) => chatRoom.toMap(),
+              ),
+          getObjectId: (e) => e.id!,
```
Comment (luke-nguyen-goldenowl) [inline lib/src/network/data/event/event_reference.dart]: change null check

File: lib/src/network/data/story/story_reference.dart [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  StoryReference()
+      : super(
+          FirebaseFirestore.instance
+              .collection('stories')
+              .withConverter<MStory>(
+                fromFirestore: (snapshot, options) => MStory.fromMap(
+                    snapshot.data() as Map<String, dynamic>, snapshot.id),
+                toFirestore: (chatRoom, _) => chatRoom.toMap(),
+              ),
+          getObjectId: (e) => e.id!,
```
Comment (luke-nguyen-goldenowl) [inline lib/src/network/data/story/story_reference.dart]: change null check

File: lib/src/network/data/story/story_reference.dart [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+          setObjectId: (e, id) => e.copyWith(id: id),
+        );
+
+  final XFirebaseStorage firebaseStorage = XFirebaseStorage();
+
+  Future<MResult<MStory>> addStory(MStory story) async {
+    try {
+      final image = await firebaseStorage.uploadImage(story.image!, "stories");
```
Comment (luke-nguyen-goldenowl) [inline lib/src/network/data/story/story_reference.dart]: change null check

File: lib/src/network/data/user_story/user_story_repository_impl.dart [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+      List<MResult<List<MStory>>> listStories = await Future.wait(futures);
+      List<MUserStory> result = [];
+      if (users.data == null) {
+        return MResult.success(result);
+      }
+      for (int i = 0; i < listStories.length; i++) {
+        if (listStories[i].isSuccess) {
+          result.add(
+              MUserStory(user: users.data![i], stories: listStories[i].data!));
```
Comment (luke-nguyen-goldenowl) [inline lib/src/network/data/user_story/user_story_repository_impl.dart]: Change null check

**Comment by Keith-Go** [comment]
resolved

**Comment by Keith-Go** [comment]
complete updated pagination for firebase

**Comment by Keith-Go** [comment]
resolved

**Comment by luke-nguyen-goldenowl** [comment]
Approved

**Pull request merged**